challenger/challenger.go

```diff
@@ -57,32 +123,146 @@ func NewChallenger(
 			c.logger.Info("New task created log received")
-			err := c.logic.ProcessNewTaskCreatedLog(newTaskCreatedLog)
+			err := c.ProcessNewTaskCreatedLog(newTaskCreatedLog)
 			if err != nil {
 				c.logger.Fatalf("Error processing NewTaskCreated log: %v", err)
 			}
 		case taskResponseLog := <-c.taskResponseChan:
 			c.logger.Info("Task response log received")
-			err := c.logic.ProcessTaskResponseLog(taskResponseLog)
+			err := c.ProcessTaskResponseLog(taskResponseLog)
 			if err != nil {
 				c.logger.Fatalf("Error processing TaskResponded log: %v", err)
 			}
 		}
 	}
 
 }
+
+func (c *Challenger[Input]) ProcessNewTaskCreatedLog(log types.Log) error {
+	var newTaskCreatedLog NewTaskCreatedEvent[Input]
+
+	err := c.taskManagerAbi.UnpackIntoInterface(&newTaskCreatedLog, "NewTaskCreated", log.Data)
+	if err != nil {
+		return fmt.Errorf("error unpacking the log: %w", err)
+	}
+
+	newTaskIndex := uint32(new(big.Int).SetBytes(log.Topics[1].Bytes()).Uint64())
+	c.tasks[newTaskIndex] = newTaskCreatedLog.InnerTask()
+
+	return nil
+}
+
+func (c *Challenger[Input]) ProcessTaskResponseLog(
+	log types.Log,
+) error {
+	var taskRespondedLog TaskRespondedEvent[Input]
+
+	err := c.taskManagerAbi.UnpackIntoInterface(&taskRespondedLog, "TaskResponded", log.Data)
+	if err != nil {
+		return fmt.Errorf("error unpacking the log: %w", err)
+	}
+
+	taskIndex := taskRespondedLog.TaskIndex()
+
+	// get the inputs necessary for raising a challenge
+	nonSigningOperatorPubKeys := c.getNonSigningOperatorPubKeys(log.TxHash)
+	taskResponseData := TaskResponseData[Input]{
+		TaskResponse:              taskRespondedLog.GetTaskResponse(),
+		TaskResponseMetadata:      taskRespondedLog.GetTaskResponseMetadata(),
+		NonSigningOperatorPubKeys: nonSigningOperatorPubKeys,
+	}
+
+	c.taskResponses[taskIndex] = taskResponseData
+
+	if task, found := c.tasks[taskIndex]; found {
+		err = c.logic.VerifyChallenge(taskIndex, task, taskResponseData)
+		if err != nil {
+			return fmt.Errorf("error verifying the challenge: %w", err)
+		}
+	}
+
+	return nil
+}
+
+func (c *Challenger[Input]) getNonSigningOperatorPubKeys(
+	transactionHash common.Hash,
+) []BN254G1Point {
+	// get the nonSignerStakesAndSignature
+	tx, _, err := c.ethClient.TransactionByHash(context.Background(), transactionHash)
+	if err != nil {
+		c.logger.Error("Error getting transaction by hash",
+			"txHash", transactionHash,
+			"err", err,
+		)
+	}
+
+	calldata := tx.Data()
+	methodSig := calldata[:4]
+	method, err := c.taskManagerAbi.MethodById(methodSig)
+	if err != nil {
+		c.logger.Error("Error getting method", "err", err)
+	}
+
+	inputs, err := method.Inputs.Unpack(calldata[4:])
+	if err != nil {
+		c.logger.Error("Error unpacking calldata", "err", err)
+	}
+
+	// Note: this implies the abi of the Task Manager contract implemented by the AVS should respect this values, or it
+	// wont work. Other solution is to receive this as parameter, but it looks more difficult than only replace this.
```

Comment:
Let's keep it as is for now. We can later take an approach similar to adc5dd25b814415825639e94e1c148e05a1a2b5e and have an abstract interface like:

```go
type ChallengerLogic[Input any, Output any] interface {
	VerifyResponse(Index, Input, Output) (bool, error)
	// this should probably include some metadata related
	// to signatures, but you get the idea.
	Challenge(index uint32, Input, Output) error
}
```

and then a low-level interface that maps 1-to-1 with the contract (and so is automatically implemented by the bindings).